presto-main/src/main/java/com/facebook/presto/sql/planner/NullabilityAnalyzer.java

```diff
@@ -128,4 +156,65 @@ protected Void visitFunctionCall(FunctionCall node, AtomicBoolean result)
             return null;
         }
     }
+
+    private static class RowExpressionVisitor
+            extends DefaultRowExpressionTraversalVisitor<AtomicBoolean>
+    {
+        private final FunctionManager functionManager;
+        private final TypeManager typeManager;
+
+        public RowExpressionVisitor(FunctionManager functionManager, TypeManager typeManager)
+        {
+            this.functionManager = functionManager;
+            this.typeManager = typeManager;
+        }
+
+        @Override
+        public Void visitCall(CallExpression call, AtomicBoolean result)
+        {
+            FunctionMetadata function = functionManager.getFunctionMetadata(call.getFunctionHandle());
+            Optional<OperatorType> operator = function.getOperatorType();
+            if (operator.isPresent()) {
+                switch (operator.get()) {
+                    case CAST: {
+                        checkArgument(call.getArguments().size() == 1);
+                        Type sourceType = call.getArguments().get(0).getType();
+                        Type targetType = call.getType();
+                        if (!typeManager.isTypeOnlyCoercion(sourceType, targetType)) {
+                            result.set(true);
+                        }
+                    }
+                    case SUBSCRIPT:
+                        result.set(true);
+                }
+            }
+            else if (!functionReturnsNullForNotNullInput(function)) {
+                // TODO: use function annotation instead of assume all function can return NULL
+                result.set(true);
+            }
+            return super.visitCall(call, result);
```

Comment:
Just realized such use cases in our codebase. This is too dangerous. It will fail if it's parent's behavior gets changed. Just write the `forEach` here. Same for `visitSpecialForm` and `InlineProjections`